src/matrixrtc/MatrixRTCSession.ts

```diff
@@ -651,32 +664,53 @@ export class MatrixRTCSession extends TypedEventEmitter<MatrixRTCSessionEvent, M
             this.emit(MatrixRTCSessionEvent.MembershipsChanged, oldMemberships, this.memberships);
         }
 
-        const isMyMembership = (m: CallMembership): boolean =>
-            m.sender === this.client.getUserId() && m.deviceId === this.client.getDeviceId();
-
         if (this.manageMediaKeys && this.isJoined() && this.makeNewKeyTimeout === undefined) {
-            const oldMebershipIds = new Set(
-                oldMemberships.filter((m) => !isMyMembership(m)).map(getParticipantIdFromMembership),
+            const oldMembershipIds = new Set(
+                oldMemberships.filter((m) => !this.isMyMembership(m)).map(getParticipantIdFromMembership),
             );
-            const newMebershipIds = new Set(
-                this.memberships.filter((m) => !isMyMembership(m)).map(getParticipantIdFromMembership),
+            const newMembershipIds = new Set(
+                this.memberships.filter((m) => !this.isMyMembership(m)).map(getParticipantIdFromMembership),
             );
 
-            const anyLeft = Array.from(oldMebershipIds).some((x) => !newMebershipIds.has(x));
-            const anyJoined = Array.from(newMebershipIds).some((x) => !oldMebershipIds.has(x));
+            const anyLeft = Array.from(oldMembershipIds).some((x) => !newMembershipIds.has(x));
+            const anyJoined = Array.from(newMembershipIds).some((x) => !oldMembershipIds.has(x));
 
             if (anyLeft) {
                 logger.debug(`Member(s) have left: queueing sender key rotation`);
                 this.makeNewKeyTimeout = setTimeout(this.onRotateKeyTimeout, MAKE_KEY_DELAY);
             } else if (anyJoined) {
                 logger.debug(`New member(s) have joined: re-sending keys`);
                 this.requestKeyEventSend();
+            } else if (this.lastMembershipFingerprints) {
+                // does it look like any of the members have updated their memberships?
+                const oldFingerprints = this.lastMembershipFingerprints;
+                this.storeLastMembershipFingerprints();
+                const newFingerprints = this.lastMembershipFingerprints;
+
+                const candidateUpdates =
+                    Array.from(oldFingerprints).some((x) => !newFingerprints.has(x)) ||
+                    Array.from(newFingerprints).some((x) => !oldFingerprints.has(x));
+                if (candidateUpdates) {
+                    logger.debug(`Member(s) have updated/reconnected: re-sending keys`);
+                    this.requestKeyEventSend();
+                }
+            } else {
+                // store the fingerprints of these latest memberships
+                this.storeLastMembershipFingerprints();
```

Comment:
Is it a problem that the fingerprints set is updated only when no one has left or joined?  Without always updating it, we'd send more key updates than necessary in some cases.  If this is a cheap operation, I guess it's fine, but maybe it's better to be correct anyways.

For example, someone joins while the fingerprints set is initialized, we'd send keys but not update the fingerprints set; then if something were to trigger another membership update without anyone having joined/left, the fingerprints set would be updated & would appear to have changed -- even though the true set of fingerprints hadn't really changed since the last join -- and we'd send keys again.

Reply:
I agree. I've [changed](https://github.com/matrix-org/matrix-js-sdk/pull/4343/commits/7a97b80861c2274026d871dc344c0c4cec59cbcf) the implementation to always recalculate the fingerprints. 👍